[Chakra v3] Fix data fetching regression with the SSR rendering (@W-18822712@)

vmarta opened the pull request:
In the chakra v3 branch, we've observed a data-fetching regression with the server rendering, where most of the queries were not being run. Only the queries in the very top level components (like App, AppConfig) would get run.

It turns out that `react-ssr-prepass` and Chakra's use of React 18's `useInsertionEffect` are not compatible. When prepass is walking through the component tree, it needs to ignore the useInsertionEffect.. similar to how it's already ignoring useEffect and useLayoutEffect.

As a workaround, we've set useInsertionEffect to a no-op from within our SDK.

There's also another similar issue where Chakra's Portal component tries to access `document` which does not exist on the server side. Now we've created our own safe version of Portal that runs only on the client side.

Remaining todos:
- [ ] add E2E test? I don't think there's such test yet for verifying the server-rendered page.

## How to test-drive the PR

1. Navigate to http://localhost:3000/?__server_only
2. Verify that the Shop Products section with its product data/images appear on the page. 
3. If yes, that means that the query to fetch data for that section was run successfully
4. Now navigate to other pages ([PLP](http://localhost:3000/global/en-GB/category/newarrivals?__server_only), [PDP](http://localhost:3000/global/en-GB/product/25697279M?color=JJ5AAXX&__server_only)) and verify that the resulting pages are very similar to the normal/client side version.

If you want to go further, you can also verify that there is no more error caught during prepass.

1. Run `npm run start:inspect` and make sure that you see the correct _server_ logs through your Chrome browser.
2. Add a debugger according to the screenshot below
3. Navigate to http://localhost:3000/?__server_only 
4. Verify that there is no error caught.

<img width="3520" height="2384" alt="Arc 2025-07-14 at 13 12 53@2x" src="https://github.com/user-attachments/assets/ad9d7dd1-9648-4818-8f1e-9f95235bf72c" />

vmarta committed:
Remove call to withLegacyGetProps

vmarta committed:
Debugging

vmarta committed:
Revert change to AppConfig

vmarta committed:
Add todos

vmarta committed:
Try commenting out the problematic Chakra hooks

Including those components that also call `useDisclosure`.

vmarta committed:
Revert "Try commenting out the problematic Chakra hooks"

This reverts commit 970b97e6f487920b9c21282f4126733efca7ed23.

vmarta committed:
Create a safe version of useDisclosure

vmarta committed:
Fix react-ssr-prepass lack of no-op for useInsertionEffect

vmarta committed:
Don't render Portal on server side

vmarta committed:
Show header again

vmarta committed:
Revert "Create a safe version of useDisclosure"

This reverts commit 669f2eb1941094e415db4936479afc10fd8970ae.

vmarta committed:
This Portal is already rendered only on client side

vmarta committed:
Render nothing on server side

vmarta committed:
Merge branch 'feature/chakra-ui-upgrade-v3' into vm/ssr-regression-fix

cc-prodsec commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/11256cff-c7cb-4b23-a1de-6fb7c286c9ca?refs=e1236fb55e3f4e809b74c2c9f3f32fb9b2012fb6&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/11256cff-c7cb-4b23-a1de-6fb7c286c9ca/license?refs=e1236fb55e3f4e809b74c2c9f3f32fb9b2012fb6&source=prChecksComment)

vmarta committed:
Clean up code

vmarta committed:
Remove unnecessary category check

vmarta committed:
Add comment

vmarta committed:
Update changelog files

kevinxh commented on packages/pwa-kit-react-sdk/src/ssr/server/react-rendering.js:
```diff
@@ -39,6 +39,12 @@ import * as errors from '../universal/errors'
+// (see https://github.com/FormidableLabs/react-ssr-prepass/issues/84)
+// - Why useInsertionEffect? Chakra v3 hooks (like useDisclosure, useBreakpointValue, useMediaQuery, useCallbackRef) rely on it for rendering optimization.
+// - Why no-op? useInsertionEffect is meant for client side only. So on the server, we wanted prepass to ignore it.
+React.useInsertionEffect = () => {} // no-op
```
Interesting fix! I'd imagine it's hard to find this out, good job!

I'm wondering if you want to move this line to near the ssr-prepass related code, for example in `packages/pwa-kit-react-sdk/src/ssr/universal/components/with-react-query/index.js`

vmarta replied:
I was thinking that we needed to set it to no-op very early on, at the start of entry point. But I can try moving it to with-react-query file and see whether the time matters or not.

unandyala replied:
Again, just trying to understand the behavior due to this change. So all the components that use useInsertionEffect will now have a slight delay (may be not noticable) for the css application?

vmarta replied:
I've just tried moving the line to the with-react-query file, but the app errors out. So I'm leaving the code change as it is now, but I've added some comment near the prepass call.

vmarta replied:
@unandyala there shouldn't be any delay nor any change in behaviour. The workaround we implemented here is only for the server side. So when useInsertionEffect is called on the _client_ side, it should behave the same as before.

🤔 Is there something else I need to consider? Let me know please, thanks.

vmarta committed:
Merge branch 'feature/chakra-ui-upgrade-v3' into vm/ssr-regression-fix

vmarta committed:
Existing tests are now passing again

vmarta commented on packages/extension-chakra-storefront/src/components/product-view/index.jsx:
```diff
@@ -40,11 +40,9 @@ const ProductViewHeader = ({
-                    <Breadcrumb categories={category} />
-                </Skeleton>
-            )}
+            <Skeleton loading={!category} minWidth={64}>
+                {category && <Breadcrumb categories={category} />}
+            </Skeleton>
```
Bonus: like what happened to the rest of the Skeletons on this page, I'm applying the same treatment here to this Skeleton. If a check is needed (in this case, checking `category`), let's make sure it's done within the Skeleton.

bendvc replied:
You might have opened a can of worms here. But do we use this component in placed where there isn't category data ever. Would this change make it so that a skeleton is showing when we don't want it to?

But this begs the question, did the old logic make any sense at all?

vmarta replied:
I've reverted the change for now :)

bendvc commented on packages/extension-chakra-storefront/CHANGELOG.md:
```diff
@@ -4,5 +4,6 @@
 - Send PWA Kit events to Data Cloud [#318] (https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2229)
 - Fix dependencies vulnerabilities [#2338](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2338)
 - Fix accessibility issues [#2375](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2375)
+- Fix data-fetching regression with SSR rendering [#2785](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2785)
```
I'm wondering if we want to be a little be more clear on this for discoverability purposes.

We can either say what you have but mention + "due to chakra v3 upgrade".

So we can say something like "Ensure client-side only API's are called on the server"

I'm also ok with leaving it as is.

bendvc commented on packages/extension-chakra-storefront/src/hooks/use-auth-modal.js:
```diff
@@ -275,6 +275,12 @@ export const AuthModal = ({
     const onBackToSignInClick = () =>
         initialView === PASSWORD_VIEW ? onClose() : setCurrentView(LOGIN_VIEW)
 
+    if (typeof window === 'undefined') {
+        // Render nothing on the server side. Otherwise, the Portal component would break SSR and its data fetching.
+        // Portal would call Ark UI's useEnvironmentContext hook, which tries to access `document` that doesn't exist on the server.
+        return
+    }
+
```
Instead of having this one off logic, should we create a <ClientOnly> component..

Also there is this new <Island clientOnly> that might fit the bill to wrap the render portion of this render. Not that I really want to solidify the Island component, but it's there.

unandyala replied:
Just trying to understand the behavior here - since the portal components are closed by default there shouldn't be any performance impact, correct?

vmarta replied:
@unandyala Yeah I think there shouldn't be any performance impact. But what performance-specific scenario or metric were you thinking of?

alexvuong replied:
@vmarta There are a few components that is using `Portal` in its render. Would this happen whenever Portal is used?

vmarta replied:
@alexvuong yeah, I'm going to replace all usage of Portal with our own version, which will render only on the client side.

bendvc commented on packages/pwa-kit-react-sdk/CHANGELOG.md:
```diff
@@ -1,4 +1,6 @@
 ## v4.0.0-extensibility-preview.5 (May 06, 2025)
+- Fix data-fetching regression with SSR rendering [#2785](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/2785)
```
Similar comment as above, but it would be nice to angle this in a positive light since "regression" almost implies we unintentionally broke something and released it and fixed it. Since we didn't release it we might be able to say "Account for rending components that use reacts useInsertionEffect hook on the server"

vmarta committed:
Revert change to the skeleton

vmarta committed:
Add a comment closer to the prepass call

The setting of useInsertionEffect to a no-op had to be done in another file.

vmarta committed:
Create a safe version of Portal

vmarta committed:
Update changelog files

vmarta committed:
Add comments to clarify

vmarta committed:
Check first whether useInsertionEffect is defined or not

vmarta committed:
Add more tests to pass test coverage

I didn't want to add test to react-rendering.test.js because my code change isn't much for testing. So I decided to add tests for another file instead.

vmarta commented on packages/pwa-kit-react-sdk/src/utils/performance.test.js:
```diff
@@ -42,4 +53,96 @@ describe('PerformanceTimer', () => {
         expect(parseFloat(timer.metrics[0].duration)).toBeGreaterThan(0)
     })
+
+    test('warns when mark name is undefined', () => {
```
To meet the test coverage, I decided to add tests for this performance util. Yes, it's not related to the PR. But there isn't much changes in the react-rendering.js file that's worthy for testing.

vmarta committed:
Add comment
